datafusion/physical-expr/src/comparison_expressions.rs

```diff
@@ -0,0 +1,268 @@
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+// Some of these functions reference the Postgres documentation
+// or implementation to ensure compatibility and are subject to
+// the Postgres license.
+
+//! Comparison expressions
+
+use arrow::array::Array;
+use arrow::datatypes::DataType;
+use arrow_ord::comparison::{gt_dyn, lt_dyn};
+use arrow_select::zip::zip;
+use datafusion_common::scalar::ScalarValue;
+use datafusion_common::{DataFusionError, Result};
+use datafusion_expr::ColumnarValue;
+
+#[derive(Debug, Clone, PartialEq)]
+enum ComparisonOperator {
+    Greatest,
+    Least,
+}
+
+macro_rules! compare_scalar_typed {
+    ($op:expr, $args:expr, $data_type:ident) => {{
+        let value = $args
+            .iter()
+            .filter_map(|scalar| match scalar {
+                ScalarValue::$data_type(v) => v.clone(),
+                _ => panic!("Impossibly got non-scalar values"),
+            })
+            .reduce(|a, b| match $op {
+                ComparisonOperator::Greatest => a.max(b),
+                ComparisonOperator::Least => a.min(b),
+            });
+        ScalarValue::$data_type(value)
+    }};
+}
+
+/// Evaluate a greatest or least function for the case when all arguments are scalars
+fn compare_scalars(
+    data_type: DataType,
+    op: ComparisonOperator,
+    args: &[ScalarValue],
+) -> ScalarValue {
+    match data_type {
+        DataType::Boolean => compare_scalar_typed!(op, args, Boolean),
+        DataType::Int8 => compare_scalar_typed!(op, args, Int8),
+        DataType::Int16 => compare_scalar_typed!(op, args, Int16),
+        DataType::Int32 => compare_scalar_typed!(op, args, Int32),
+        DataType::Int64 => compare_scalar_typed!(op, args, Int64),
+        DataType::UInt8 => compare_scalar_typed!(op, args, UInt8),
+        DataType::UInt16 => compare_scalar_typed!(op, args, UInt16),
+        DataType::UInt32 => compare_scalar_typed!(op, args, UInt32),
+        DataType::UInt64 => compare_scalar_typed!(op, args, UInt64),
+        DataType::Float32 => compare_scalar_typed!(op, args, Float32),
+        DataType::Float64 => compare_scalar_typed!(op, args, Float64),
+        DataType::Utf8 => compare_scalar_typed!(op, args, Utf8),
+        DataType::LargeUtf8 => compare_scalar_typed!(op, args, LargeUtf8),
+        _ => panic!("Unsupported data type for comparison: {:?}", data_type),
+    }
+}
+
+/// Evaluate a greatest or least function
+fn compare(op: ComparisonOperator, args: &[ColumnarValue]) -> Result<ColumnarValue> {
```

Comment:
I think this functionality is basically the same as the Min/Max accumulators (e.g. https://docs.rs/datafusion/latest/datafusion/physical_plan/expressions/struct.MinAccumulator.html)